web/src/message_events.js

```diff
@@ -364,6 +374,8 @@ export function update_messages(events) {
                     message_id: msg.id,
                 });
             }
+            // Add data after updating stream_id, topic and edit_history fields.
+            all_messages_data.add_messages(event_messages);
```

Comment:
I am not sure whether we should add all messages here together or one by one. The comment in the below code results in this confusion -
```python
for (const msg of messages) {
    // Put messages in correct order on either side of the
    // message list.  This code path assumes that messages
    // is a (1) sorted, and (2) consecutive block of
    // messages that belong in this message list; those
    // facts should be ensured by the caller.
    if (this.empty() || msg.id > this.last_including_muted().id) {
        bottom_messages.push(msg);
    } else if (msg.id < this.first_including_muted().id) {
        top_messages.push(msg);
    } else {
        interior_messages.push(msg);
    }
}

if (interior_messages.length > 0) {
    interior_messages = this.add_anywhere(interior_messages);
}

if (top_messages.length > 0) {
    top_messages = this.prepend(top_messages);
}

if (bottom_messages.length > 0) {
    bottom_messages = this.append(bottom_messages);
}
```

The comment mentions that "messages should be a consecutive block of messages", but from what I understand from the code we add the messages at the top, bottom and in middle by sorting as per the messages already in the list. So, do we still want the consecutive blocks of messags?

Reply:
It's important that `all_messages_data` be a contiguous range, because of the logic in `narrow.js` that uses it as a cache to populate message lists locally before talking to the server.

Probably what this means is we need some sort of logic for this live-update code of the form "Process messages within the range between the first and last message IDs in all_messages_data."

Reply:
Just to be clear the above code is in general for all `MessageListData` objects.

Reply:
> One more thing I noticed, that we do not remove message from message_store on deleting them. While we do remove them from all_messages_data.

That sounds like a bug.

Reply:
> Just to be clear the code in the above comment is in general for all MessageListData objects.

I guess actually this change could be wrong for a normal MessageListData object too. Like, the thing that is correct is "please re-process these messages that existing in the MessageListData object". What is not correct is adding messages, even if they exist in the message store, to a MessageListData object without the explicit guarantees that the `message_fetch` code path provides that those messages are within the contiguous interval of message IDs in the `MessageListData`. 

The data structure is a range of messages, say it's currently [100, 200], and if you added message ID 53 to it because of a live-update code path like this, then if you scroll to the top of that view, now we'll ask the server for the messages starting from 52 to append to the top, and messages 54-99 will be incorrectly never fetched or displayed.

(It's extra bad if this happens in `all_messages_data`, because that'll potentially corrupt every narrow you visit after that point, but it applies to any narrow)

Reply:
Currently, we add messages to `all_messages_data` and `message_store` together, and if we would fix the case of removing the message from `message_store` on deleting the messages, then we would be removing them from both the data structures together as well. So, I think this should make sure that we are not adding any message that did not previously existed in `all_messages_data` if we are already removing the messages that are not present in `message_store`. 

Or may be as you suggested we can check the range of messages already present in `all_messages_data` before removing them to make sure we add only once that were present previously.

Reply:
Anyways, I can first try implementing the bulk operation mentioned in TODO.